File: leetcode/balanced_binary_tree/BalancedBinaryTree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode *root) {
        if (!root) {
            return true; 
        }
        
        int leftHeight = height(root->left); 
        int rightHeight = height(root->right); 
        
        if (abs(leftHeight - rightHeight) > 1) {
            return false; 
        }
        
        return isBalanced(root->left) || isBalanced(root->right); 
    }
    
    int height(TreeNode * node) {
        if (!node) {
            return -1; 
        }
        
        return max(height(node->left), height(node->right)) + 1; 
    }
};

File: lintcode/balanced_binary_tree.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: True if this Binary tree is Balanced, or false.
     */
    public boolean isBalanced(TreeNode root) {
        // write your code here
        if (root == null) {
            return true; 
        }
        
        int leftHeight = getHeight(root.left);
        int rightHeight = getHeight(root.right);
        int diff = Math.abs(leftHeight - rightHeight);
        
        if (diff > 1) {
            return false; 
        }
        
        return isBalanced(root.left) && isBalanced(root.right);
    }
    
    private int getHeight(TreeNode node) {
        if (node == null) {
            return 0; 
        }
        
        return Math.max(getHeight(node.left), getHeight(node.right)) + 1;
    }
}
